File: php/src/detectors/XssDetector.php
<?php

namespace FireWTWall\Detectors;

class XssDetector
{
    private static array $rules = [
        ['name' => 'xss-script-tag',       'severity' => 'critical', 'pattern' => '/<\s*script[\s>\/]/i'],
        ['name' => 'xss-javascript-proto', 'severity' => 'critical', 'pattern' => '/javascript\s*:/i'],
        ['name' => 'xss-vbscript-proto',   'severity' => 'critical', 'pattern' => '/vbscript\s*:/i'],
        ['name' => 'xss-data-uri',         'severity' => 'critical', 'pattern' => '/data\s*:\s*text\/html/i'],
        ['name' => 'xss-event-handler',    'severity' => 'high',     'pattern' => '/\bon\w+\s*=/i'],
        ['name' => 'xss-iframe',           'severity' => 'high',     'pattern' => '/<\s*iframe[\s>\/]/i'],
        ['name' => 'xss-object-embed',     'severity' => 'high',     'pattern' => '/<\s*(?:object|embed)[\s>\/]/i'],
        ['name' => 'xss-svg',             'severity' => 'high',     'pattern' => '/<\s*svg[\s>\/]/i'],
        ['name' => 'xss-link-meta',        'severity' => 'medium',   'pattern' => '/<\s*(?:link|meta)[\s>\/]/i'],
        ['name' => 'xss-expression',       'severity' => 'high',     'pattern' => '/expression\s*\(/i'],
        ['name' => 'xss-img-src',          'severity' => 'medium',   'pattern' => '/<\s*img[^>]+src\s*=/i'],
        ['name' => 'xss-srcdoc',           'severity' => 'high',     'pattern' => '/srcdoc\s*=/i'],
        ['name' => 'xss-base-href',        'severity' => 'medium',   'pattern' => '/<\s*base[\s>]/i'],
        ['name' => 'xss-form-action',      'severity' => 'high',     'pattern' => '/<\s*form[^>]+action\s*=\s*[\'"]?javascript/i'],
        ['name' => 'xss-dom-write',        'severity' => 'high',     'pattern' => '/document\s*\.\s*(?:write|writeln)\s*\(/i'],
        ['name' => 'xss-inner-html',       'severity' => 'high',     'pattern' => '/\.innerHTML\s*=/i'],
        ['name' => 'xss-angularjs-bind',   'severity' => 'high',     'pattern' => '/\{\{.*\}\}/'],
    ];

    public static function scan(array $sources): ?array
    {
        foreach ($sources as $label => $values) {
            $result = self::scanValues($values, $label);
            if ($result !== null) return $result;
        }
        return null;
    }

    private static function scanValues($data, string $label): ?array
    {
        if (is_string($data)) {
            return self::matchString($data, $label);
        }
        if (is_array($data)) {
            foreach ($data as $value) {
                $r = self::scanValues($value, $label);
                if ($r !== null) return $r;
            }
        }
        return null;
    }

    private static function matchString(string $value, string $label): ?array
    {
        foreach (self::$rules as $rule) {
            if (preg_match($rule['pattern'], $value, $m)) {
                return [
                    'rule'     => $rule['name'],
                    'severity' => $rule['severity'],
                    'matched'  => substr($m[0], 0, 120),
                    'source'   => $label,
                ];
            }
        }
        return null;
    }
}

File: aspnet/src/detectors/XssDetector.cs
using System.Collections.Generic;
using System.Text.RegularExpressions;

namespace FireWTWall.Detectors
{
    public static class XssDetector
    {
        private static readonly (string Name, string Severity, Regex Pattern)[] Rules =
        {
            ("xss-script-tag",       "critical", new Regex(@"<\s*script[\s>/]",                              RegexOptions.IgnoreCase | RegexOptions.Compiled)),
            ("xss-javascript-proto", "critical", new Regex(@"javascript\s*:",                                RegexOptions.IgnoreCase | RegexOptions.Compiled)),
            ("xss-vbscript-proto",   "critical", new Regex(@"vbscript\s*:",                                  RegexOptions.IgnoreCase | RegexOptions.Compiled)),
            ("xss-data-uri",         "critical", new Regex(@"data\s*:\s*text/html",                          RegexOptions.IgnoreCase | RegexOptions.Compiled)),
            ("xss-event-handler",    "high",     new Regex(@"\bon\w+\s*=",                                   RegexOptions.IgnoreCase | RegexOptions.Compiled)),
            ("xss-iframe",           "high",     new Regex(@"<\s*iframe[\s>/]",                              RegexOptions.IgnoreCase | RegexOptions.Compiled)),
            ("xss-object-embed",     "high",     new Regex(@"<\s*(?:object|embed)[\s>/]",                   RegexOptions.IgnoreCase | RegexOptions.Compiled)),
            ("xss-svg",              "high",     new Regex(@"<\s*svg[\s>/]",                                 RegexOptions.IgnoreCase | RegexOptions.Compiled)),
            ("xss-link-meta",        "medium",   new Regex(@"<\s*(?:link|meta)[\s>/]",                      RegexOptions.IgnoreCase | RegexOptions.Compiled)),
            ("xss-expression",       "high",     new Regex(@"expression\s*\(",                               RegexOptions.IgnoreCase | RegexOptions.Compiled)),
            ("xss-img-src",          "medium",   new Regex(@"<\s*img[^>]+src\s*=",                          RegexOptions.IgnoreCase | RegexOptions.Compiled)),
            ("xss-srcdoc",           "high",     new Regex(@"srcdoc\s*=",                                   RegexOptions.IgnoreCase | RegexOptions.Compiled)),
            ("xss-base-href",        "medium",   new Regex(@"<\s*base[\s>]",                                RegexOptions.IgnoreCase | RegexOptions.Compiled)),
            ("xss-form-action",      "high",     new Regex(@"<\s*form[^>]+action\s*=\s*['""]?javascript",  RegexOptions.IgnoreCase | RegexOptions.Compiled)),
            ("xss-dom-write",        "high",     new Regex(@"document\s*\.\s*(?:write|writeln)\s*\(",       RegexOptions.IgnoreCase | RegexOptions.Compiled)),
            ("xss-inner-html",       "high",     new Regex(@"\.innerHTML\s*=",                              RegexOptions.IgnoreCase | RegexOptions.Compiled)),
            ("xss-angularjs-bind",   "high",     new Regex(@"\{\{.*\}\}",                                   RegexOptions.Compiled)),
            ("xss-css-import",       "high",     new Regex(@"@import\s+url\s*\(",                           RegexOptions.IgnoreCase | RegexOptions.Compiled)),
            ("xss-moz-binding",      "high",     new Regex(@"-moz-binding\s*:",                             RegexOptions.IgnoreCase | RegexOptions.Compiled)),
            ("xss-meta-refresh",     "high",     new Regex(@"<meta[^>]+http-equiv\s*=\s*[""']?refresh",    RegexOptions.IgnoreCase | RegexOptions.Compiled)),
            ("xss-link-import",      "high",     new Regex(@"<link[^>]+rel\s*=\s*[""']?import",            RegexOptions.IgnoreCase | RegexOptions.Compiled)),
            ("xss-svg-animate",      "medium",   new Regex(@"<animate[^>]+attributeName",                   RegexOptions.IgnoreCase | RegexOptions.Compiled)),
        };

        public static DetectorResult Scan(Dictionary<string, string> sources)
        {
            foreach (var kv in sources)
            {
                var r = MatchString(kv.Value, kv.Key);
                if (r != null) return r;
            }
            return null;
        }

        private static DetectorResult MatchString(string value, string label)
        {
            foreach (var rule in Rules)
            {
                Match m = rule.Pattern.Match(value);
                if (m.Success)
                    return new DetectorResult(rule.Name, rule.Severity,
                        m.Value.Length > 120 ? m.Value.Substring(0, 120) : m.Value, label);
            }
            return null;
        }
    }
}
